subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaLibraryPlugin.java

```diff
@@ -0,0 +1,88 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.api.plugins;
+
+import com.google.common.collect.ImmutableMap;
+import org.gradle.api.Action;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.ConfigurationContainer;
+import org.gradle.api.internal.artifacts.configurations.ConfigurationInternal;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.compile.JavaCompile;
+
+import static org.gradle.api.plugins.JavaPlugin.USAGE;
+
+/**
+ * <p>A {@link Plugin} which extends the capabilities of the {@link JavaPlugin Java plugin} by cleanly separating
+ * the API and implementation dependencies of a library.</p>
+ */
+public class JavaLibraryPlugin implements Plugin<Project> {
+
+    @Override
+    public void apply(final Project project) {
+        project.getPluginManager().apply(JavaPlugin.class);
+
+        JavaPluginConvention convention = (JavaPluginConvention) project.getConvention().getPlugins().get("java");
+        final ConfigurationContainer configurations = project.getConfigurations();
+        addApiToMainSourceSet(project, convention, configurations);
+    }
+
+    private void addApiToMainSourceSet(final Project project, JavaPluginConvention convention, final ConfigurationContainer configurations) {
+        convention.getSourceSets().getByName("main", new Action<SourceSet>() {
+            @Override
+            public void execute(SourceSet sourceSet) {
+                defineApiConfigurationsForSourceSet(sourceSet, configurations);
+                JavaCompile javaCompile = (JavaCompile) project.getTasks().findByName(sourceSet.getCompileJavaTaskName());
+                project.getArtifacts().add(sourceSet.getApiCompileConfigurationName(), ImmutableMap.of(
+                    "file", javaCompile.getDestinationDir(),
+                    "builtBy", javaCompile));
+            }
+        });
+    }
+
+    private void defineApiConfigurationsForSourceSet(SourceSet sourceSet, ConfigurationContainer configurations) {
+        Configuration apiConfiguration = configurations.maybeCreate(sourceSet.getApiConfigurationName());
+        apiConfiguration.setVisible(false);
+        apiConfiguration.setDescription("API dependencies for " + sourceSet + ".");
+        apiConfiguration.setCanBeResolved(false);
+        apiConfiguration.setCanBeConsumed(false);
+        apiConfiguration.setTransitive(false);
+
+        Configuration apiCompileConfiguration = configurations.maybeCreate(sourceSet.getApiCompileConfigurationName());
+        apiCompileConfiguration.setVisible(false);
+        apiCompileConfiguration.setDescription("API compile classpath for " + sourceSet + ".");
+        apiCompileConfiguration.setCanBeResolved(true);
```

Comment:
Yes, it is only meant for consumers, but for debugging it can be practical to allow resolution (typically print the compile classpath). That's not critical, though.